compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp

```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
+    RewriterBase &rewriter, scf::ForallOp forallOp, int64_t flatSize,
+    Value flatId, bool syncAfterDistribute,
+    std::optional<transform::TransformOpInterface> transformOp,
+    const ArrayRef<DeviceMappingAttrInterface> &threadMappingAttributes) {
+  // Step 0. Target-specific verifications. There is no good place to anchor
+  // those right now: the ForallOp is target-independent and the
+  // transform op does not apply to individual ForallOp.
+  auto failureHelper =
+      [&](const Twine &message) -> DiagnosedSilenceableFailure {
+    if (transformOp.has_value()) {
+      return transformOp->emitSilenceableError() << message;
+    }
+    return emitDefiniteFailure(forallOp, message);
+  };
+  Location loc = forallOp->getLoc();
+  if (!forallOp.isNormalized())
+    return failureHelper("unsupported non-normalized loops");
+  if (forallOp.getNumResults() > 0)
+    return failureHelper("only bufferized scf.forall lowers to gpu.thread_id");
+  if (forallOp.getRank() > 3)
+    return failureHelper(
+        "scf.forall with rank > 3 does not lower to gpu.thread_id");
+  if (llvm::any_of(forallOp.getMixedUpperBound(), [](OpFoldResult ofr) {
+        return !getConstantIntValue(ofr).has_value();
+      })) {
+    return failureHelper("unsupported dynamic blockdim size");
+  }
+  if (!forallOp.getMapping().has_value())
+    return failureHelper("mapping must be present");
+  SmallVector<Attribute> threadMapping =
+      llvm::to_vector(forallOp.getMapping()->getValue());
+
+  // Step 1. Complete the threadMapping to a full mapping (with 1s) if
+  // necessary.
+  SmallVector<Value> numThreads = forallOp.getUpperBound(rewriter);
+  // Ensure we have 3 block sizes, one for each id.
+  Value one;
+  for (auto attr : threadMappingAttributes) {
+    if (std::find(threadMapping.begin(), threadMapping.end(), attr) ==
+        threadMapping.end()) {
+      threadMapping.push_back(attr);
+      one = one ? one : rewriter.create<arith::ConstantIndexOp>(loc, 1);
+      numThreads.push_back(one);
+    }
+  }
+
+  // Step 2. sort the values by the corresponding DeviceMappingAttrInterface.
+  auto comparator = [&](DeviceMappingAttrInterface a,
+                        DeviceMappingAttrInterface b) -> bool {
+    return a.getMappingId() < b.getMappingId();
+  };
+  SmallVector<Value> blockDimValues = scf::ForallOp::getValuesSortedByKey(
+      threadMapping, numThreads, comparator);
+  SmallVector<int64_t> blockDims =
+      llvm::to_vector(llvm::map_range(blockDimValues, [](Value v) {
+        return v.getDefiningOp<arith::ConstantIndexOp>().value();
+      }));
+
+  // Step 3. Create the gpu.thread ops and map the induction variables to the
+  // newly created ops.
+  Value flatThreadId = flatId;
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  SmallVector<Value> threadOpsUpdated;
+  int64_t flatblockDim = 1;
+  for (int64_t dim : blockDims) {
+    flatblockDim *= dim;
+  }
+  int64_t dimUsed = 1;
+  for (int64_t dim : blockDims) {
```

Comment:
All this code looks like common linearization/delinearization in a basis for which we have facilities upstream (https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L96), why did you need to roll your own?

If you need more functionality please adapt.